src/app/test-results/test-results-list/test-results-list.component.ts

```diff
@@ -0,0 +1,54 @@
+import { Component, Input, OnInit } from '@angular/core';
+import { TestResultService } from 'api-client';
+import { TestResultSummary } from 'src/app/models/test-result-summary.model';
+import { TestResultDetail } from 'src/app/models/test-result-detail.model';
+
+const NO_SELECTION = -1;
+@Component({
+  selector: 'wh-test-results-list',
+  templateUrl: './test-results-list.component.html',
+})
+export class TestResultsListComponent implements OnInit {
+  @Input() buildId: number;
+  summaries: TestResultSummary[];
+  details: TestResultDetail[];
+  failedDetails: TestResultDetail[];
+  selectedSummaryId: number = NO_SELECTION;
+  selectedDetailId: number = NO_SELECTION;
+
+  constructor(
+    private testResultService: TestResultService) { }
+
+  ngOnInit(): void {
+    this.testResultService
+      .buildBuildidTestResultSummaryGet(this.buildId)
+      .subscribe(summaries => {
+        this.summaries = summaries;
+      });
+  }
+
+  toggleDetails(summary: TestResultSummary): void {
+    if (this.selectedSummaryId === summary.testResultSummaryId) {
+      this.selectedSummaryId = NO_SELECTION;
+      this.selectedDetailId = NO_SELECTION;
+    } else {
+      this.selectedSummaryId = summary.testResultSummaryId;
+      this.selectedDetailId = NO_SELECTION;
+    }
+
+    this.testResultService
+      .buildBuildidTestResultSummaryArtifactIdDetailGet(summary.buildId, summary.artifactId)
+      .subscribe(details => {
+        this.details = details;
+        this.failedDetails = this.details.filter(detail => detail.status === 'Failed');
+      });
```

Comment:
This implementation lacks:
- Caching the test results details. If you close and expand the same artifact, it fetches new data even though it just fetched it.
- Doesn't allow multiple artifacts to be expanded at the same time.

Suggest splitting up this component into smaller sub-components. Dealing with special values on a per-list-item-basis is much easier if each list item is its own component and it just deals with its own special values.

Such as how we extracted `project-list-item` out of `project-list` to reduce the complexity and to let each `project-list-item` to hold their own state such as if the project is refreshed or not. I think it applies we here as well, where you could change to having each `test-results-list-item` component deal with if they're expanded or not, and if they've already fetched the data or not.

Abuse the fact that Angular has a component modelled architecture :)

Reply:
The recent changes should have made it fulfill those things now.